provider/pkg/provider/handlers.go

```diff
@@ -0,0 +1,98 @@
+// Copyright 2021, Pulumi Corporation.
+
+package provider
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/imdario/mergo"
+	"github.com/pkg/errors"
+	"github.com/pulumi/pulumi-google-native/provider/pkg/resources"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/resource"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/util/logging"
+	rpc "github.com/pulumi/pulumi/sdk/v3/proto/go"
+)
+
+type updateHandler interface {
+	Update(inputs resource.PropertyMap, oldState resource.PropertyMap) (map[string]interface{}, error)
+}
+
+type updateHandlerFunc func(inputs resource.PropertyMap, oldState resource.PropertyMap) (map[string]interface{}, error)
+
+func (u updateHandlerFunc) Update(inputs resource.PropertyMap, oldState resource.PropertyMap) (map[string]interface{}, error) {
+	return u(inputs, oldState)
+}
+
+func (p *googleCloudProvider) updateHandler(resourceTok string, req *rpc.UpdateRequest) updateHandler {
+	logging.V(9).Infof("UpdateHandler for %q", resourceTok)
+	switch resourceTok {
+	case "google-native:container/v1:Cluster", "google-native:container/v1beta1:Cluster":
```

Comment:
Consider an approach similar to Azure Native does: add an extra property to resource metadata to define the update style, set it while generating metadata, and rely on it here rather than resource names. It reduces duplication of resource names and makes this an explicit property of a resource that can be tracked over time.